### PR TITLE
riotboot: add riotboot_flashwrite module

### PR DESCRIPTION
**Pull request by kaspar030**
### Contribution description

This PR adds a module to riotboot that allows writing a firmware image to an inactive firmware slot, along with a simple test application.

On its own, it doesn't do any validity checks of the to-be-flashed image. The module is supposed to be a building block used by higher level modules, e.g., for updating via validated suit manifests.

### Testing procedure

Testing is a bit manual at the moment... Please see the README.md that comes with the test application.

### Issues/PRs references

~~Depends on #11180 for the testing instructions to work.~~ (merged)

### CODE REVIEW

**Comment by kaspar030** [comment]
- rebased (with #11180 merged)

File: sys/include/riotboot/flashwrite.h [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+
+/**
+ * @brief   Initialize firmware update (raw version)
+ *
+ * Allows setting the initial offset to @p offset, which would mean that the
+ * first bytes passed in via @ref riotboot_flashwrite_putbytes() will be
+ * written to (slot_start + offset).
+ *
+ * @note offset *should* be <= FLASHPAGE_SIZE, otherwise the results are
+ *       undefined.
```
[the remark this answers is not in the document]
Comment (kaspar030) [reply sys/include/riotboot/flashwrite.h]: yup, added.

File: sys/include/riotboot/flashwrite.h [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    int target_slot;                        /**< update targets this slot     */
+    size_t offset;                          /**< update is at this position   */
+    unsigned flashpage;                     /**< update is at this flashpage  */
+    uint8_t flashpage_buf[FLASHPAGE_SIZE];  /**< flash writing buffer         */
+} riotboot_flashwrite_t;
+
+/**
+ * @brief Amount of bytes to skip at initial write of first page
+ */
+#define RIOTBOOT_FLASHWRITE_SKIPLEN 4
```
[the remark this answers is not in the document]
Comment (kaspar030) [reply sys/include/riotboot/flashwrite.h]: ok

File: tests/riotboot_flashwrite/README.md [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+Please see the README of examples/nanocoap_server for instructions on how to
+set up a network for testing.
+
+# How to test
+
+First, compile and flash with riotboot enabled:
+
+    $ BOARD=<board> make riotboot/flash
+
+Configm it booted from slot 0, then recompile in order to get an image for the
```
Comment (bergzand) [inline tests/riotboot_flashwrite/README.md]: This sentence doesn't parse for me :-/

Comment (kaspar030) [reply tests/riotboot_flashwrite/README.md]: cleared that up!

Comment (bergzand) [reply tests/riotboot_flashwrite/README.md]: Thanks!

File: sys/include/riotboot/flashwrite.h [hunk 4]
```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright (C) 2018 Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     sys_firmware
```
Comment (bergzand) [inline sys/include/riotboot/flashwrite.h]: `@ingroup sys_riotboot`?

File: sys/riotboot/flashwrite.c [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+ * Copyright (C) 2019 Inria
+ *               2019 Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     sys_riotboot
```
Comment (bergzand) [inline sys/riotboot/flashwrite.c]: `@ingroup sys_riotboot`?

File: sys/include/riotboot/flashwrite.h [hunk 6]
```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright (C) 2018 Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    sys_riotboot_flashwrite riotboot flash writer
```
Comment (bergzand) [inline sys/include/riotboot/flashwrite.h]: *R*iotboot?

Comment (kaspar030) [reply sys/include/riotboot/flashwrite.h]: IMO all the riotboot module documentation needs work. I suggest we postpone that to a later PR.

Comment (bergzand) [reply sys/include/riotboot/flashwrite.h]: :+1:

**Comment by kaspar030** [comment]
&go.

**Comment by cladmi** [comment]
This pull request broke this that was working:

```
FEATURES_REQUIRED+=riotboot BOARD=samr21-xpro make -C examples/hello-world/ riotboot/flash
```

**Comment by cladmi** [comment]
I disagree, the fix from https://github.com/RIOT-OS/RIOT/pull/11203 do not fix the issues.

This pull request changed the API without mentioning in the title or giving an adequate testing description.

The third commit https://github.com/RIOT-OS/RIOT/pull/11181/commits/a15f07b04bb3153c69c7dfcb3aa9b5cb23724c07 is added without justification and changes the definition scope of the makefiles variables and made some non namespaced macros globally defined.
This may be correct, but not without a specific discussion and a justification other than "move slot variables". It should have be in its own pull request. Also, if the variables are now parsed by Makefile.include why keep the export ?

If it was required for https://github.com/RIOT-OS/RIOT/pull/11181/commits/ae35860510915461544384ccde57d9455095f18d why is it defined after the commit introducing it ?

And the fix proposed in #11203  now makes that having `FEATURES_REQUIRED+=riotboot` define global non namespaced macros for no reasons.


This should really have been split to a separate pull request with dedicated testing procedure and justification as this pull request did not gave any and did not mention it either.

**Pull request merged**